Simplify replicate logic

zdevito opened the pull request:
This moves the logic around for handling parameters in replicate so that it has fewer conditionals for script modules.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36188 Fix long line splitting issue in python_print
* #36179 Use Function to implement fork.
* **#36174 Simplify replicate logic**
* #36168 _requires_grad -> requires_grad

Differential Revision: [D20903301](https://our.internmc.facebook.com/intern/diff/D20903301)

zdevito committed:
Simplify replicate logic

[ghstack-poisoned]

zdevito committed:
Update on "Simplify replicate logic"

This moves the logic around for handling parameters in replicate so that it has fewer conditionals for script modules.


Differential Revision: [D20903301](https://our.internmc.facebook.com/intern/diff/D20903301)

[ghstack-poisoned]

jamesr66a commented:
Seems to break on DataParallel https://dr.pytorch.org/api/view-log-full?build_id=112111303

ngimel commented on torch/nn/modules/module.py:
```diff
@@ -1179,7 +1179,10 @@ def __dir__(self):
+
+        # replicas do not have parameters themselves, the replicas reference the original
+        # module.
+        replica._parameters = []
```
_parameters is an OrderedDict, not a list.

zdevito committed:
Update on "Simplify replicate logic"

This moves the logic around for handling parameters in replicate so that it has fewer conditionals for script modules.


Differential Revision: [D20903301](https://our.internmc.facebook.com/intern/diff/D20903301)

[ghstack-poisoned]

zdevito committed:
Update on "Simplify replicate logic"

This moves the logic around for handling parameters in replicate so that it has fewer conditionals for script modules.


Differential Revision: [D20903301](https://our.internmc.facebook.com/intern/diff/D20903301)

[ghstack-poisoned]

zdevito committed:
Update on "Simplify replicate logic"

This moves the logic around for handling parameters in replicate so that it has fewer conditionals for script modules.


Differential Revision: [D20903301](https://our.internmc.facebook.com/intern/diff/D20903301)

[ghstack-poisoned]

zdevito committed:
Update on "Simplify replicate logic"

This moves the logic around for handling parameters in replicate so that it has fewer conditionals for script modules.


Differential Revision: [D20903301](https://our.internmc.facebook.com/intern/diff/D20903301)

[ghstack-poisoned]
